### DISCUSSION TITLE
Question about  tcl.vim

### DISCUSSION THREAD

**Discussion by avl42** [vim/vim#16045] [Q&A]
https://github.com/vim/vim/blame/7724d621f21ff6599ba49f11401ddb45ff321099/runtime/syntax/tcl.vim#L130

This and the following line seem to be there for already 20years, but only now I found them, and don't understand them.

In particular, they spoil the highlighting for Tcl-scripts few lines below any occurrence of command "variable".

Tcl-command "variable" just takes a list of arguments being variable names and initialization values alternately.
e.g.:  variable foo "forty two" bar 42 snafu ;#  all get linked to namespace; foo and bar also get overwritten, snafu not.

I commented them out in my local installation of vim, and all seems fine, now I'd like to know what could have been the original reason for them...   I'm not yet claiming they're wrong, but maybe someone still remembers the motives for adding.

**Comment by chrisbra** [comment]
you really should try to check back with the actual maintainer of the file.

**Comment by rhysd** [comment]
Would you change the category to Q&A? This thread is unexpectedly broadcasted to feeds because of 'Announcements' category.

**Reply by chrisbra** [reply]
sorry, din't think about it. And I have no idea, why discussions converted from issues appear at the 'Announcements' category

**Comment by dkearns** [comment]
@avl42 If you don't have any luck with the maintainer can you please post a specific example exhibiting this behaviour?  I had a very quick look at some files and I don't see it.

**Reply by avl42** [reply]
The following example is taken from  https://github.com/avl42/temper/blob/Tcl/temper.tcl  (with many lines removed)

```
namespace eval USBList {
   proc device_info {dirname} {
      variable BLANKDEV; variable ATTRS
      
      set device $BLANKDEV
      return $device
   }
}
```
It looks like this with the tcl.vim as shipped:
![grafik](https://github.com/user-attachments/assets/96aab95d-7836-4a9a-a58b-3acdaaf9221e)

If I comment out (edit: "uncomment" was the wrong term here) 
the referenced lines (130,131) in my local machine's tcl.vim, then it looks like this:
![grafik](https://github.com/user-attachments/assets/0d7d44c3-7653-4b79-a74e-ea1d6e1cad80)

The main difference is, that the first command (ditto with a comment) after a "variable" command
is not syntax-highlighted at all:  here the "set", in other places it is a comment.  Maybe this depends
on the particular arguments to the variable command, these here each just give a variable to be tied
to the namespace.

PS: now that I see the screenshotlets in the preview: the second one has caught a bit of the
 hollow cursor - that's of course not tcl.vim's fault ;-)

**Comment by dkearns** [comment]
I haven't spent much time on this and my Tcl is _very_ rusty but it seems the minimal case for your bug is the following (whitespace required before `variable`):

```tcl
    variable BLANKDEV; variable ATTRS
    set device $BLANKDEV
```

The `variable` region isn't terminated by the `;` as it really should be and contains another `variable` region which prevents the first from terminating as it consumes the EOL.  Adding `keepend` as mentioned in the comment should fix that but it's an odd way to define the regions.

Presumably these were created to offer more precise matching of `variable` commands but they don't seem to make much practical difference, at first glance, as the contained items are matched at top level anyway.

There's other problems though, the first region definition never matches at all as `variable` is declared as a keyword earlier in the file and has priority, tail comments aren't matched, and the blank line `skip` options don't make sense as that's not valid syntax.

I'm also not sure why they're defined as `tcltkCommand` rather than `tclCommand` like the earlier definition in the file.

Did you have any luck contacting the maintainer?
